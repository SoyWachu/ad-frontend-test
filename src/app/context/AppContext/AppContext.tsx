"use client";
import { createContext, useContext } from "react";
import type { AppState, AppAction } from "@/app/reducers/appReducer";

export const AppContext = createContext<{
  state: AppState;
  dispatch: React.Dispatch<AppAction>;
  fetchGames: (
    page?: number,
    genre?: string,
    append?: boolean
  ) => Promise<void>;
} | null>(null);

export function useApp() {
  const context = useContext(AppContext);
  if (!context) throw new Error("useApp debe usarse dentro de AppProvider");
  return context;
}
