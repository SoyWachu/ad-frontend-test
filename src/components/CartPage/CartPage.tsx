"use client";
import React from "react";
import { useApp } from "@/app/context/AppContext/AppContext";
import Link from "next/link";

export default function CartPage() {
  const { state, dispatch } = useApp();

  const handleRemove = (id: string) => {
    dispatch({ type: "REMOVE_ITEM", payload: id });
  };

  const totalItems = state.cart.reduce(
    (acc, item) => acc + (item.quantity || 1),
    0
  );
  const totalPrice = state.cart.reduce(
    (acc, item) => acc + item.price * (item.quantity || 1),
    0
  );

  return (
    <div className="flex flex-col gap-6 p-6 px-5 sm:px-5 lg:px-32">
      <div className="flex justify-between items-center">
        <Link
          href="/"
          className="text-gray-700 hover:text-gray-900 font-semibold transition-colors"
        >
          <span>← Back to Catalog</span>
        </Link>
      </div>

      <p className="text-lg font-bold sm:text-4xl">Your Cart</p>
      <p className="text-xl">
        {totalItems} item{totalItems !== 1 ? "s" : ""}
      </p>

      <div className="flex flex-col lg:flex-row gap-8">
        <div data-testid="cart-product-list" className="flex-1 space-y-6">
          {state.cart.length === 0 ? (
            <p>Your cart is empty.</p>
          ) : (
            state.cart.map((item) => (
              <div
                key={item.id}
                data-testid="cart-item"
                className="flex flex-col sm:flex-row gap-4 border-b p-4 transition hover:bg-gray-50 w-[95%]"
              >
                <div className="flex justify-between items-start w-full sm:w-auto">
                  <img
                    src={item.imageUrl}
                    alt={item.title}
                    className="w-full sm:w-72 h-54 sm:h-40 object-cover"
                  />
                  <button
                    aria-label={`remove ${item.title}`}
                    onClick={() => handleRemove(item.id)}
                    className="text-gray-400 hover:text-gray-600 transition-colors sm:hidden ml-2"
                  >
                    ✕
                  </button>
                </div>

                <div className="flex-1 flex flex-col justify-between mt-2 sm:mt-0">
                  <div className="flex justify-between items-center">
                    <p className="text-xs text-gray-400 uppercase">
                      {item.genre}
                    </p>
                    <button
                      aria-label={`remove ${item.title}`}
                      onClick={() => handleRemove(item.id)}
                      className="text-gray-400 hover:text-gray-600 transition-colors hidden sm:block"
                    >
                      ✕
                    </button>
                  </div>

                  <div className="mt-1">
                    <h2 className="font-semibold">{item.title}</h2>
                    <p className="text-gray-400 text-sm mt-1 line-clamp-2">
                      {item.description}
                    </p>
                    {item.quantity && item.quantity > 1 && (
                      <p className="text-gray-500 text-sm mt-1">
                        Quantity: {item.quantity}
                      </p>
                    )}
                  </div>

                  <div className="flex justify-end mt-2">
                    <p className="font-semibold">
                      ${(item.price * (item.quantity || 1)).toFixed(2)}
                    </p>
                  </div>
                </div>
              </div>
            ))
          )}
        </div>

        <div className="w-full lg:w-[40%] flex flex-col">
          <div
            data-testid="order-summary"
            className="bg-white border rounded-lg shadow p-6"
          >
            <h2 className="font-semibold text-lg mb-1">Order Summary</h2>
            <p className="text-gray-500 text-sm mb-6">
              {totalItems} item{totalItems > 1 ? "s" : ""}
            </p>

            <div className="space-y-3 mb-6">
              {state.cart.map((item) => (
                <div key={item.id} className="flex justify-between text-sm">
                  <span>{item.title}</span>
                  <span>${(item.price * (item.quantity || 1)).toFixed(2)}</span>
                </div>
              ))}
            </div>

            <hr className="border-gray-300 mb-6 w-full" />

            <div className="flex justify-between font-semibold">
              <span>Order Total</span>
              <span>${totalPrice.toFixed(2)}</span>
            </div>
          </div>

          <button className="mt-4 w-full bg-[#585660] text-white py-3 rounded-lg hover:bg-gray-800 transition">
            Checkout
          </button>
        </div>
      </div>
    </div>
  );
}
