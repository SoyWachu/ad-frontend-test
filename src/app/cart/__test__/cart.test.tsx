import { render, screen, fireEvent, within } from "@testing-library/react";
import CartPage from "@/components/CartPage/CartPage";
import React from "react";
import { AppContext } from "@/app/context/AppContext/AppContext";

const baseMockState = {
  games: [],
  totalPages: 1,
  currentPage: 1,
  loading: false,
  availableFilters: [],
};

const mockStateWithItems = {
  ...baseMockState,
  cart: [
    {
      id: "1",
      title: "Game 1",
      genre: "Action",
      description: "Awesome game",
      price: 100,
      quantity: 2,
      imageUrl: "/game1.png",
    },
    {
      id: "2",
      title: "Game 2",
      genre: "RPG",
      description: "Another great game",
      price: 30,
      quantity: 1,
      imageUrl: "/game2.png",
    },
  ],
};

const mockStateEmpty = {
  ...baseMockState,
  cart: [],
};

const mockDispatch = jest.fn();
const mockFetchGames = jest.fn();

describe("CartPage", () => {
  beforeEach(() => {
    mockDispatch.mockClear();
  });

  it("renders cart items correctly", () => {
    render(
      <AppContext.Provider
        value={{
          state: mockStateWithItems,
          dispatch: mockDispatch,
          fetchGames: mockFetchGames,
        }}
      >
        <CartPage />
      </AppContext.Provider>
    );

    const productList = screen.getByTestId("cart-product-list");
    const productItems = within(productList).getAllByTestId("cart-item");

    expect(within(productItems[0]).getByText("Game 1")).toBeInTheDocument();
    expect(within(productItems[1]).getByText("Game 2")).toBeInTheDocument();
    expect(
      within(productItems[0]).getByText(/Quantity: 2/i)
    ).toBeInTheDocument();
  });

  it("calculates total items and total price correctly", () => {
    render(
      <AppContext.Provider
        value={{
          state: mockStateWithItems,
          dispatch: mockDispatch,
          fetchGames: mockFetchGames,
        }}
      >
        <CartPage />
      </AppContext.Provider>
    );

    const orderSummary = screen.getByTestId("order-summary");
    expect(within(orderSummary).getByText("$230.00")).toBeInTheDocument();
    expect(within(orderSummary).getByText(/3 items/i)).toBeInTheDocument();
  });

  it("removes item from cart when clicking remove button", () => {
    render(
      <AppContext.Provider
        value={{
          state: mockStateWithItems,
          dispatch: mockDispatch,
          fetchGames: mockFetchGames,
        }}
      >
        <CartPage />
      </AppContext.Provider>
    );

    const productList = screen.getByTestId("cart-product-list");
    const productItems = within(productList).getAllByTestId("cart-item");

    const removeButtons = within(productItems[0]).getAllByLabelText(
      "remove Game 1"
    );

    fireEvent.click(removeButtons[0]);

    expect(mockDispatch).toHaveBeenCalledWith({
      type: "REMOVE_ITEM",
      payload: "1",
    });
  });

  it("shows empty cart message when cart is empty", () => {
    render(
      <AppContext.Provider
        value={{
          state: mockStateEmpty,
          dispatch: mockDispatch,
          fetchGames: mockFetchGames,
        }}
      >
        <CartPage />
      </AppContext.Provider>
    );

    expect(screen.getByText("Your cart is empty.")).toBeInTheDocument();
    expect(screen.getByText(/0 items/i)).toBeInTheDocument();
    expect(screen.getByText("$0.00")).toBeInTheDocument();
  });

  it("renders back to catalog link correctly", () => {
    render(
      <AppContext.Provider
        value={{
          state: mockStateWithItems,
          dispatch: mockDispatch,
          fetchGames: mockFetchGames,
        }}
      >
        <CartPage />
      </AppContext.Provider>
    );

    expect(screen.getByText(/← Back to Catalog/i)).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <AppContext.Provider
        value={{
          state: mockStateWithItems,
          dispatch: mockDispatch,
          fetchGames: mockFetchGames,
        }}
      >
        <CartPage />
      </AppContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
