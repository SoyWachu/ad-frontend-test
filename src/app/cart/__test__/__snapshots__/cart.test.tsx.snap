// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CartPage matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-6 p-6 px-5 sm:px-5 lg:px-32"
  >
    <div
      class="flex justify-between items-center"
    >
      <a
        href="/"
      >
        <span>
          ← Back to Catalog
        </span>
      </a>
    </div>
    <p
      class="text-lg font-bold sm:text-4xl"
    >
      Your Cart
    </p>
    <p
      class="text-xl"
    >
      3
       item
      s
    </p>
    <div
      class="flex flex-col lg:flex-row gap-8"
    >
      <div
        class="flex-1 space-y-6"
        data-testid="cart-product-list"
      >
        <div
          class="flex flex-col sm:flex-row gap-4 border-b p-4 transition hover:bg-gray-50 w-[95%]"
          data-testid="cart-item"
        >
          <div
            class="flex justify-between items-start w-full sm:w-auto"
          >
            <img
              alt="Game 1"
              class="w-full sm:w-72 h-54 sm:h-40 object-cover"
              src="/game1.png"
            />
            <button
              aria-label="remove Game 1"
              class="text-gray-400 hover:text-gray-600 transition-colors sm:hidden ml-2"
            >
              ✕
            </button>
          </div>
          <div
            class="flex-1 flex flex-col justify-between mt-2 sm:mt-0"
          >
            <div
              class="flex justify-between items-center"
            >
              <p
                class="text-xs text-gray-400 uppercase"
              >
                Action
              </p>
              <button
                aria-label="remove Game 1"
                class="text-gray-400 hover:text-gray-600 transition-colors hidden sm:block"
              >
                ✕
              </button>
            </div>
            <div
              class="mt-1"
            >
              <h2
                class="font-semibold"
              >
                Game 1
              </h2>
              <p
                class="text-gray-400 text-sm mt-1 line-clamp-2"
              >
                Awesome game
              </p>
              <p
                class="text-gray-500 text-sm mt-1"
              >
                Quantity: 
                2
              </p>
            </div>
            <div
              class="flex justify-end mt-2"
            >
              <p
                class="font-semibold"
              >
                $
                200.00
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col sm:flex-row gap-4 border-b p-4 transition hover:bg-gray-50 w-[95%]"
          data-testid="cart-item"
        >
          <div
            class="flex justify-between items-start w-full sm:w-auto"
          >
            <img
              alt="Game 2"
              class="w-full sm:w-72 h-54 sm:h-40 object-cover"
              src="/game2.png"
            />
            <button
              aria-label="remove Game 2"
              class="text-gray-400 hover:text-gray-600 transition-colors sm:hidden ml-2"
            >
              ✕
            </button>
          </div>
          <div
            class="flex-1 flex flex-col justify-between mt-2 sm:mt-0"
          >
            <div
              class="flex justify-between items-center"
            >
              <p
                class="text-xs text-gray-400 uppercase"
              >
                RPG
              </p>
              <button
                aria-label="remove Game 2"
                class="text-gray-400 hover:text-gray-600 transition-colors hidden sm:block"
              >
                ✕
              </button>
            </div>
            <div
              class="mt-1"
            >
              <h2
                class="font-semibold"
              >
                Game 2
              </h2>
              <p
                class="text-gray-400 text-sm mt-1 line-clamp-2"
              >
                Another great game
              </p>
            </div>
            <div
              class="flex justify-end mt-2"
            >
              <p
                class="font-semibold"
              >
                $
                30.00
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="w-full lg:w-[40%] flex flex-col"
      >
        <div
          class="bg-white border rounded-lg shadow p-6"
          data-testid="order-summary"
        >
          <h2
            class="font-semibold text-lg mb-1"
          >
            Order Summary
          </h2>
          <p
            class="text-gray-500 text-sm mb-6"
          >
            3
             item
            s
          </p>
          <div
            class="space-y-3 mb-6"
          >
            <div
              class="flex justify-between text-sm"
            >
              <span>
                Game 1
              </span>
              <span>
                $
                200.00
              </span>
            </div>
            <div
              class="flex justify-between text-sm"
            >
              <span>
                Game 2
              </span>
              <span>
                $
                30.00
              </span>
            </div>
          </div>
          <hr
            class="border-gray-300 mb-6 w-full"
          />
          <div
            class="flex justify-between font-semibold"
          >
            <span>
              Order Total
            </span>
            <span>
              $
              230.00
            </span>
          </div>
        </div>
        <button
          class="mt-4 w-full bg-[#585660] text-white py-3 rounded-lg hover:bg-gray-800 transition"
        >
          Checkout
        </button>
      </div>
    </div>
  </div>
</div>
`;
